src/wire/dhcp.rs

```diff
@@ -0,0 +1,632 @@
+        field.copy_from_slice(value.as_bytes());
+    }
+
+    pub fn set_server_ip(&mut self, value: Ipv4Address) {
+        let field = &mut self.buffer.as_mut()[field::SIADDR];
+        field.copy_from_slice(value.as_bytes());
+    }
+
+    pub fn set_relay_agent_ip(&mut self, value: Ipv4Address) {
+        let field = &mut self.buffer.as_mut()[field::GIADDR];
+        field.copy_from_slice(value.as_bytes());
+    }
+
+    pub fn set_broadcast_flag(&mut self, value: bool) {
+        let field = &mut self.buffer.as_mut()[field::FLAGS];
+        NetworkEndian::write_u16(field, if value { 1 } else { 0 });
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + AsMut<[u8]> + ?Sized> Packet<&'a mut T> {
+    /// Return a pointer to the options.
+    #[inline]
+    pub fn options_mut(&mut self) -> Result<&mut [u8]> {
+        let data = self.buffer.as_mut();
+        data.get_mut(field::OPTIONS).ok_or(Error::Truncated)
+    }
+}
+
+
+enum_with_unknown! {
+    /// The possible opcodes/message types of a DHCP packet.
+    pub enum MessageType(u8) {
+        Discover = 1,
+        Offer = 2,
+        Request = 3,
+        Decline = 4,
+        Ack = 5,
+        Nak = 6,
+        Release = 7,
+        Inform = 8,
+    }
+}
+
+impl MessageType {
+    fn opcode(&self) -> Result<u8> {
+        match *self {
+            MessageType::Discover | MessageType::Inform | MessageType::Request |
+                MessageType::Decline | MessageType::Release => Ok(1),
+            MessageType::Offer | MessageType::Ack | MessageType::Nak => Ok(2),
+            MessageType::Unknown(_) => Err(Error::Malformed),
+        }
+    }
+}
+
+enum_with_unknown! {
+    /// The possible hardware address types of a DHCP packet.
+    pub enum HardwareAddressType(u8) {
+        Ethernet = 1,
+        ExperimentalEthernet = 2,
+        AmateurRadio = 3,
+        ProNet = 4,
+        Chaos = 5,
+        Ieee802 = 6,
+        Arcnet = 7,
+        Hyperchannel = 8,
+        Lanstar = 9,
+    }
+}
+
+/// A representation of a single DHCP option.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum DhcpOption<'a> {
+    EndOfList,
+    MessageType(MessageType),
+    RequestedIp(Ipv4Address),
+    ServerIdentifier(Ipv4Address),
+    Other { kind: u8, data: &'a [u8] }
+}
+
+impl<'a> DhcpOption<'a> {
+    pub fn parse(buffer: &'a [u8]) -> Result<(&'a [u8], DhcpOption<'a>)> {
+        // See https://tools.ietf.org/html/rfc2132 for all possible DHCP options.
+
+        let (length, option);
+        match *buffer.get(0).ok_or(Error::Truncated)? {
+            field::OPT_END => {
+                length = 1;
+                option = DhcpOption::EndOfList;
+            }
+            field::OPT_PAD => {
+                unimplemented!();
+                // “The pad option can be used to cause subsequent fields to align on word
+                // boundaries.”
+                // TODO: what does this mean? Align the next field or align all following fields?
```

Comment:
No, this just means that servers may insert pad options if they want, and clients should ignore them. It's for convenience of server implementers.